atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1370,104 +1365,75 @@ private void commitWrites(TransactionService transactionService) {
+                // we risk another transaction starting at a timestamp after our commit timestamp not seeing our writes.
+                timedAndTraced("commitWrite", () -> keyValueService.multiPut(writesByTable, getStartTimestamp()));
+
+                // Now that all writes are done, get the commit timestamp
+                // We must do this before we check that our locks are still valid to ensure that other transactions that
+                // will hold these locks are sure to have start timestamps after our commit timestamp.
+                long commitTimestamp = timedAndTraced("getCommitTimestamp", timelockService::getFreshTimestamp);
+                commitTsForScrubbing = commitTimestamp;
+
+                // Punch on commit so that if hard delete is the only thing happening on a system,
+                // we won't block forever waiting for the unreadable timestamp to advance past the
+                // scrub timestamp (same as the hard delete transaction's start timestamp).
+                // May not need to be here specifically, but this is a very cheap operation - scheduling another thread
+                // might well cost more.
+                timedAndTraced("microsForPunch", () -> cleaner.punch(commitTimestamp));
+
+                // Serializable transactions need to check their reads haven't changed, by reading again at
+                // commitTs + 1. This must happen before the lock check for thorough tables, because the lock check
+                // verifies the immutable timestamp hasn't moved forward - thorough sweep might sweep a conflict out
+                // from underneath us.
+                timedAndTraced("readWriteConflictCheck",
+                        () -> throwIfReadWriteConflictForSerializable(commitTimestamp));
+
+                // Verify that our locks and pre-commit conditions are still valid before we actually commit;
+                // this throwIfPreCommitRequirementsNotMet is required by the transaction protocol for correctness.
+                timedAndTraced("preCommitLockCheck", () -> throwIfImmutableTsOrCommitLocksExpired(commitLocksToken));
+                timedAndTraced("userPreCommitCondition", () -> throwIfPreCommitConditionInvalid(commitTimestamp));
+
+                timedAndTraced("commitPutCommitTs",
+                        () -> putCommitTimestamp(commitTimestamp, commitLocksToken, transactionService));
+
+                long microsSinceCreation = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis() - timeCreated);
+                getTimer("commitTotalTimeSinceTxCreation").update(microsSinceCreation, TimeUnit.MICROSECONDS);
+                getHistogram(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_BYTES_WRITTEN).update(byteCount.get());
+            } finally {
+                timedAndTraced("postCommitUnlock",
+                        () -> timelockService.tryUnlock(ImmutableSet.of(commitLocksToken)));
+            }
+        });
+    }
```

Comment:
I've verified correspondence between the original flow of steps and what we have here, which looks right.